WIP: PKCS#11 v3.0 NIST 800-108 KBKDF

cipherboy opened the pull request:
With messy commits.

# Errata updated Nov 7th, 2019

Current PKCS#11 v3.0 errata that I've addressed in this implementation (per @rjrelyea, submitted to PKCS#11 committee): 

 1. PKCS#11 v3.0 [Cryptographic Token Interface Current Mechanisms](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc437440586) specification specifies a `CK_SP800_108_COUNTER`, while the [`pkcs11t.h`](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/csprd01/include/pkcs11-v3.0/pkcs11t.h) from PKCS#11 v3.0 Cryptographic [Token Interface Base Specification](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html) specifies `CK_SP800_108_OPTIONAL_COUNTER`. 
    - I've provided both, making `CK_SP800_108_COUNTER` define to `CK_SP800_108_OPTIONAL_COUNTER`
 2. In both `CK_SP800_108_KDF_PARAMS` and `CK_SP800_108_FEEDBACK_KDF_PARAMS`, `pAdditionalDerivedKeys` is a single `CK_DERIVED_KEY`, when it should be a `CK_DERIVED_KEY_PTR`, since it is interpreted as an array of `CK_DERIVED_KEY`s. This appears in both the CTICM and `pkcs11t.h`.
    - I've fixed it in both places.
 3. PKCS#11 v3.0 [Cryptographic Token Interface Current Mechanisms](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc437440586) specification specifies `CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS` and `CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS` methods (and defines what they do), but the [`pkcs11t.h`](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/csprd01/include/pkcs11-v3.0/pkcs11t.h) from PKCS#11 v3.0 Cryptographic [Token Interface Base Specification](https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html) doesn't assign them constant values.
    - I've fixed it by assigning 1 to the former, 2 to the latter:
      ```c
      /* ERRATA: PKCS#11 v3.0 Cryptographic Token Interface Current Mechanisms
       * defines that these constants exist, but doesn't specify values. pkcs11t.h
       * from PKCS#11 v3.0 Cryptographic Token Interface Base Specification doesn't
       * define these constants either. */
      #define CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS     0x00000001UL
      #define CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS 0x00000002UL
      ```
 4. ~For all three sections, PKCS#11 v3.0 says: about `CK_SP800_108_DKM_LENGTH`, "This data field type is optional.", however no default method is chosen in the later section on [deriving additional keys](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc8118477). So either PKCS#11 v3.0 should define a default, or make the field required when `pAdditionalDerivedKeys` is non-empty (and make it explicitly ignored otherwise).~
    - ~Additionally, without this field, there's no way to indicate the endianness of L parameter, so perhaps it would be better to make this a required field.~ 
    - This was my misunderstanding of the spec. I incorrectly conflated `DKM` with the _additional_ derived keys, not the main derived key, in part because the struct is both an indicator for *where* the length field is and *how* it is calculated with additional derived keys. I think it is fine as-is. 
5. `CK_HANDLE_INVALID` isn't defined anywhere in the PKCS#11 v3.0 header files. This is referenced from [PKCS#11 v3.0 section 2.42.6 Deriving Additional Keys](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc8118477).
    - This should be `CK_INVALID_HANDLE`. 

Unaddressed erratas (per @rjrelyea, not submitted to committee) :

 1. [NIST SP800-108](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf) section leaves `h` as a "fixed value". Note that in this case "fixed value" means "decided upon ahead of time", not a fundamental constant (as the other most common "fixed value" is the size of the counter in bits). Most use cases of KBKDF (say [RFC 6803](https://tools.ietf.org/html/rfc6803) and SCP03) use the full size of the PRF. However, in SCP03's instance, they explicitly clarify it as "used with full 16 byte output length" -- to me this reads like the authors interpreted SP800-108 as allowing the PRF to be truncated (say taking 128 bits at a time out of SHA-2-256). Regardless of whether or not this is addressed, it does seem relevant to pass PRF Mechanism data from the caller if necessary. Though, current mechanisms don't need it.
 2. In Feedback mode, the field "`K(i-1)`" is denotated as type `CK_SP800_108_ITERATION_VARIABLE` and requires a `CK_SP800_108_COUNTER_FORMAT` instance. PKCS#11 v3.0 mentions at the end of the description, "Exact formatting of the counter value is defined by the `CK_SP800_108_COUNTER_FORMAT` structure." -- however, outside of possible truncation of the PRF output (which isn't mentioned but could be argued to be `h`), which could be specified in `ulWidthInBits`, there's no need for both items of the `CK_SP800_108_COUNTER_FORMAT` struct, as SP800-108 generally defines truncation to be the left most bits. That type is thus misleading and using a separate type would be cleaner. This also applies to Pipeline mode and "`A(i)`". Additionally, SP800-108 states "optionally, using  a  counter  as  the  iteration  variable(s). " -- so calling "`K(i-1)`"/"`A(i)`" that value is further misleading. 
    - Concretely, I'd introduce a `PREVIOUS_RESULT` type, make it optionally contain a length (which could be different than `h` discussed in the previous errata), and drop either `CK_SP800_108_ITERATION_VARIABLE` or `CK_SP800_108_COUNTER` (as only one is ever necessary) type identifier. Probably the former. I'd then introduce another identifier called `CK_SP800_108_PREVIOUS_RESULT` and have it be a separate parameter type, `CK_SP800_108_PRF_RESULT_FORMAT`, which contains a single `CK_ULONG`, the number of bits from the previous PRF invocation to take (which is optionally zero). Allow this field to be NULL to signify taking all the bits since that is the implied default and this field, when present, would be mandatory.
3. ~It really isn't clear how `pAdditionalDerivedKeys` are supposed to work.~ You have an `CK_ATTRIBUTE_PTR` array of attributes. [PKCS#11 v3.0 section 2.42.6 Deriving Additional Keys](https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc8118477) talks about layout of these additional keys. ~But as far as I can tell, nowhere does it say which attributes are used for placement. There's `CKA_VALUE_LEN` to be used to determine the _size_ of the additional key to derive. But nothing to determine placement / offset in the key stream.~ 
    - I'd consider introducing `CKA_NSS_KDF_OFFSET` to represent the offset in the KDF output stream this key exists at. It is of type `CK_ULONG`.
    - The way they're _supposed_ to work is that each key gets filled in order (function-passed key first, then additional derived keys), where each key needs to be filled starting at a "fresh" PRF invocation. I'm wondering if any consumer of NIST SP800-108 actually uses the KDF this way. It seems really convoluted. Something like `CKA_NSS_KDF_OFFSET` makes a lot more sense, with no restrictions on PRF output.
    - As far as I can tell, these restrictions don't come from NIST SP800-108 itself. So I'm confused where these restrictions came from.

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_gtest.gyp

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/manifest.mn

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

rjrelyea commented on gtests/pk11_gtest/pk11_kbkdf.cc:
```diff
@@ -0,0 +1,72 @@
+      dataParams,
+      0,       /* no additional derived keys */
+      NULL     /* no additional derived keys */
+    };
```
Should have at least one hmac test as well.

cipherboy replied:
My intent is to add everything from NIST's published [CAVP tests](https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Key-Derivation) for this, since this implementation _should_ be able to support it all.

cipherboy replied:
@rjrelyea I've since added all CAVP tests for Counter mode, sans the TDES / 2TDES tests. So this includes:

 - AES-128
 - AES-192
 - AES-256
 - HMAC-SHA1
 - HMAC-SHA224
 - HMAC-SHA256
 - HMAC-SHA384
 - HMAC-SHA512

So there should be no surprises here during FIPS certification of this. OTOH, there are a _lot_ of tests:

 - Fixed data before ("counter after" / `AFTER_FIXED`)
 - Fixed data before and after ("counter middle" / `MIDDLE_FIXED`)
 - Fixed data after ("counter before" / `BEFORE_FIXED`)

With `r` bitlen between:

 - 8
 - 16
 - 24
 - 32

Each of these has several tests for each of the different ciphers. My thought was to prune this to say, 1-2 of each (`r` bit length, MAC function, output length) combination. Thoughts?

rjrelyea commented on lib/softoken/pkcs11c.c:
```diff
@@ -8231,6 +8231,54 @@ NSC_DeriveKey(CK_SESSION_HANDLE hSession,
+            }
+
+            crv = kbkdf_Pipeline((CK_SP800_108_KDF_PARAMS *)pMechanism->pParameter, sourceKey, &key);
+            break;
```
This exactly how it should be added to Derive (as you can tell the Derive switch is quite large, so all new drive mechanisms should be added just like this).

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,427 @@
+        /* params->prfType != CKM_SHA3_384_HMAC && */
+        /* params->prfType != CKM_SHA3_512_HMAC && */
+        /* params->prfType != CKM_3DES_CMAC && */
+        params->prfType != CKM_AES_CMAC
```
I'd like it better if this had a more generic "is this CKM_ a valid mac". There may be something like that for ike. At the very least, test to see if it's an hmac so that we don't have to go over the entire set of derive functions when we add a new mac (particularly a new hmac).

cipherboy replied:
Good point. Where does "glue" code such as this live? Should I just put it in `lib/softoken/pkcs11c.c`?

The implementation in `sftkike.c` doesn't seem to validate if something is a valid (H)MAC and doesn't currently support CMAC (but does support AES XCBC). `sftk_map_hmac_to_hash` could be reused for the latter HMAC->Hash mapping. There's a comment there that says: 

```c
/* Map the mechanism to the underlying hash. If the type is not a hash
 * or HMAC, return HASH_AlgNULL. This can happen legitimately if
 * we are doing AES XCBC */
```

So we could probably extend it and turn it into a proper mapper / validator, at least for HMACs.

rjrelyea replied:
yes, I think this can move to probably pkcs11u.c and add it to the pkcs11i.h header. A mapping of hmac to Hash_alg mechanism is generally useful, Something that gets a general mac interface is better. Even with the map_hmac_to_hash function, just checking that the mechansim is an hmac, then handling the cmac separately is a win.

rjrelyea replied:
How sftk_ike handles the parameter tests:

```c
static CK_RV
prf_setup(prfContext *context, CK_MECHANISM_TYPE mech)
{
    context->hashType = sftk_map_hmac_to_hash(mech);
    context->hashObj = NULL;
    context->hmac = NULL;
    context->aes = NULL;
    if (context->hashType != HASH_AlgNULL) {
        context->hashObj = HASH_GetRawHashObject(context->hashType);
        if (context->hashObj == NULL) {
            return CKR_GENERAL_ERROR;
        }
        return CKR_OK;
    } else if (mech == CKM_AES_XCBC_MAC) {
        return CKR_OK;
    }
    return CKR_MECHANISM_PARAM_INVALID;
}
```

rjrelyea replied:
We could take the prf_ functions in ike and move them to pkcs11i.h and then add cmac to them. That addes cmac to ike and aes_xcbc to sp100-108.

cipherboy replied:
I don't personally think there's much value in extending SP800-108 to use AES-XCBC as a mechanism type. We certainly could, and that's PKCS#11's prerogative. But XCBC (CBC-MAC / OMAC1) isn't FIPS standardized (whether that's good or bad is subjective). 

Regardless, its use seems limited to [IPSec](https://tools.ietf.org/html/rfc3566) and other major protocols like TLS seem to have largely ignored it.

SP800-108 is kinda also in that bucket. But both SCP03 and Kerberos have found use for it, and neither with AES-XCBC.

I think I'd rather go with a (HMAC+CMAC) union and leave AES-XCBC as an outlier for IKE to implement. 

My 2c.

rjrelyea replied:
I don't think we need to support AES_XCBC in our SP800-108 implementation. I just wanted in the generic implementation so we could use if for IKE.

rjrelyea replied:
In any case you should at least use some helper function to handle if the hmac is supported. You don't want to keep updating this code every time we add a new hash function.

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,427 @@
+    case CKM_SHA512_HMAC:
+        hashObj = HASH_GetRawHashObject(HASH_AlgSHA512);
+        ctx->prf_size = SHA512_LENGTH;
+        goto hmac;
```
There has to be a 'mechanism to hashObj function available for softoken at this level'. I can see cmac could have some diffs, at least at startup, but hmac should all be automatic.

cipherboy replied:
@rjrelyea I was actually hoping to avoid this all together but I wasn't yet quite sure how so I went with the way I knew would work. :)

My idea was to use the existing PKCS#11 layer to handle this for us. Rather than holding the `MACCtx ctx` object everywhere (and having to handle the different HMAC/CMAC initializations/update/finalization code), keep only a `CK_MECHANISM_PTR pMechanism` (and at most a `CK_SESSION_HANDLE hSession`) to be able to call `NSC_SignInit`/`NSC_SignUpdate`/`NSC_Sign`.

The questions I had left unresolved were the blockers:

 - Can I reuse `hSession` from the parent (NIST 800-108 KBKDF) context or do I have to create a new one for HMAC/CMAC?
 - How do I construct `CK_MECHANISM_PTR`?

Anyhow, I went with this current approach so I could work towards having a valid Counter implementation while I figured the answers to those questions out. 

Thoughts?

cipherboy replied:
(And to answer the question: `ike` has a `sftk_map_hmac_to_hash` function we could probably reuse once we add a header definition).

rjrelyea replied:
So that's a bit tricky. sftk_InitCBCMac() sort of does this. The problem is technically the session could already have a Sign session going on it (not common, but possible), so reusing that session would be iffy. We could create a new session. You'll also need to create specific keys (since the input key is likely modified before using in the hmac). sftk_InitCBCMac does this as well (It creates a temp key used for the underlying AES operation), That key is  freed before the Init returns. We have to remember to free any of these things we create in the operation.

In ike, I created a generic mac interface that kindof handled both hmac (which is pretty easy) and aes_xcbc. Maybe we should create a completely generic version that handles all the macing. That might be too much for this patch, but it might make a good future bug.

bob

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,427 @@
+        break;
+    case CKM_AES_CMAC:
+        CMAC_Destroy(ctx->mac.cmac, PR_TRUE);
+    default:
```
If you add the destroy function as a pointer in startup, you won't have to repeat the mechanism switch here.

cipherboy replied:
Yeah, that's in an unpushed change. :)

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,427 @@
+/* [ section: parameter helpers ] */
+static CK_RV kbkdf_CalculateParameters() {
+    return CKR_OK;
+}
```
Obviously this needs to be added.

rjrelyea reviewed (commented):
Obviously there is more to do in this patch, but I thought I'd give you some general comments.

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP do-build.sh

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP do-test.sh

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy commented:
- PKCS#11 v3.0 NIST SP800-108 documentation: https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html#_Toc437440585
- PKCS#11 v3.0 general documentation: https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html
- NIST SP800-108 publication: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-108.pdf

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_counterbeforefixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_counterbeforefixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_counterafterfixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_countermiddlefixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy commented on do-build.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
@rjrelyea You can ignore these scripts. They won't make it into the final NSS check in for obvious reasons.

cipherboy commented on do-test.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
@rjrelyea You can ignore these scripts. They won't make it into the final NSS check in for obvious reasons.

cipherboy commented on gtests/pk11_gtest/pk11_kbkdf.cc:
```diff
@@ -0,0 +1,224 @@
+    }
+    assert(result != NULL);
+
+    assert(PK11_ExtractKeyValue(result.get()) == SECSuccess);
```
@rjrelyea Would it be preferable to make these tests pass in FIPS mode, by using the key (of type `CKM_SHA512_HMAC`, per above) and making it compute say, `HMAC(key, NULL)` and compare the signatures?

I do notice this particular function `PK11_ExtractKeyValue` is used elsewhere in the test suite, e.g., `ssl_gtest/tls_hkdf_unittest.cc`, `pk11_gtest/pk11_pbkdf2_unittest.cc`, and `pk11_gtest/pk11_encrypt_derive_unittest.cc`.

cipherboy replied:
Never mind. Per earlier email discussions, this isn't possible since we're calling import key here.

cipherboy commented on gtests/pk11_gtest/pk11_kbkdf.cc:
```diff
@@ -0,0 +1,224 @@
+    return result;
+  }
+
+  void RunCounterBeforeFixedTest(CK_MECHANISM_TYPE prf_mech, uint32_t counter_bitlen, uint32_t output_bitlen, uint8_t *key, uint32_t key_len, uint8_t *fixed_input, uint32_t fixed_input_len, uint8_t *expected) {
```
I know I need to clean up these tests and simplify them.

cipherboy commented on lib/softoken/pkcs11c.c:
```diff
@@ -8279,6 +8279,54 @@ NSC_DeriveKey(CK_SESSION_HANDLE hSession,
+                crv = CKR_MECHANISM_PARAM_INVALID;
+                break;
+            }
+            if (sftk_isTrue(key, CKA_TOKEN)) {
```
Should I move these two checks (`sftk_isTrue` and `sftk_DeriveSensitiveCheck`) into `kbkdf_Counter`?

cipherboy commented on lib/softoken/pkcs11c.c:
```diff
@@ -8279,6 +8279,54 @@ NSC_DeriveKey(CK_SESSION_HANDLE hSession,
                                   sourceKey, key);
             break;
 
+        case CKM_SP800_108_COUNTER_KDF:
```
If we want to save space in a cluttered function, we could make this:

```c
case CKM_SP800_108_COUNTER_KDF:
case CKM_SP800_108_FEEDBACK_KDF:
case CKM_SP800_108_DOUBLE_PIPELINE_KDF
    crv = kbkdf_Dispatch(...);
    break;
```

Or some such, instead of adding all these lines here. That'd also allow us to unify handling of the additional safety checks in a single, common place for these functions. Thoughts?

cipherboy reviewed (commented):
@rjrelyea :

 - ~I still need to simplify and clean up this diff. There's some debugging statements I left in.~
 - ~I haven't implemented feedback or pipeline modes yet.~
 - ~I haven't implemented additional key material.~
 - ~I haven't cleaned up the tests yet.~ -- Does NSS have CAVP parsing here?
 - I haven't tested additional key material yet.
 - I haven't implemented writing data to non-keys, like the TLS PRF.
 - ~I haven't simplified and unified the MAC constructs.~

But it does a basic counter mode!

My thoughts are:

 1. Propose a unified MAC construct as separate PR, then submit just that to NSS upstream. It'd probably be in the FreeBL layer, use that in the IKE code. 
    - Do we want AES-XCBC MAC as part of that? I'm leaning towards "no" since it isn't a primitive in FreeBL. Thoughts and preferences? 
 2. Finish TLS PRF-like data creation.
 3. Finish additional key material creation.
 4. Finish pipeline and feedback mode.
 5. Clean up and polish.
 6. Propose to NSS upstreams.

cipherboy reviewed (commented):
.

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy commented on lib/softoken/pkcs11c.c:
```diff
@@ -2004,111 +2004,42 @@ sftk_HMACCmp(CK_ULONG *copyLen, unsigned char *sig, unsigned int sigLen,
 }
 
 /*
- * common HMAC initalization routine
```
@rjrelyea Review these changes.

cipherboy commented on lib/softoken/pkcs11i.h:
```diff
@@ -798,6 +801,55 @@ sftk_TLSPRFInit(SFTKSessionContext *context,
                 HASH_HashType hash_alg,
                 unsigned int out_len);
 
+/* MAC-agnostic PKCS#11 implementation */
```
@rjrelyea Review this.

cipherboy commented on lib/softoken/sftkhmac.c:
```diff
@@ -15,6 +15,8 @@ static HASH_HashType
 MACMechanismToHash(CK_MECHANISM_TYPE mech)
 {
     switch (mech) {
+        case CKM_MD2_HMAC:
```
@rjrelyea Review this.

cipherboy commented on lib/softoken/sftkhmac.c:
```diff
@@ -188,3 +214,239 @@ sftk_MACConstantTime_DestroyContext(void *pctx, PRBool free)
+    return ret;
+}
+
+CK_RV sftk_MAC_InitRaw(SftkMACCtx *ctx, CK_MECHANISM_TYPE mech, const unsigned char *key, unsigned int key_len, PRBool isFIPS) {
```
This is what lets it work for the IKE case. We could add XCBC MAC into this implementation if you wanted to, but I'd prefer to make that a separate patch once KBKDF is proposed upstream.

rjrelyea replied:
OK, That sounds fine.

cipherboy reviewed (commented):
@rjrelyea I've asked for feedback on some particular places. I'd like to propose this as a MAC cleanup patch upstream. Thoughts? As time permits later, I can work on including XCBC MAC into this MAC structure.

rjrelyea commented on lib/softoken/sftkhmac.c:
```diff
@@ -188,3 +214,239 @@ sftk_MACConstantTime_DestroyContext(void *pctx, PRBool free)
+
+    ret = sftk_MAC_InitRaw(ctx, mech,
+                           (const unsigned char *)keyval->attrib.pValue,
+                           keyval->attrib.ulValueLen, isFIPS);
```
Good to separate out sftk_MAC_InitRaw(). There are likely lots of places in softoken where we have the key bits and not a key handle.

rjrelyea commented on lib/softoken/sftkhmac.c:
```diff
@@ -188,3 +214,239 @@ sftk_MACConstantTime_DestroyContext(void *pctx, PRBool free)
+    return ret;
+}
+
+CK_RV sftk_MAC_Reset(SftkMACCtx *ctx) {
```
Yes, this is good. There are cases where we need to execute the same mac multiple times using the same keys. It comes up in ike as well.

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+    SFTKSession *session = NULL;
+
+    /* TODO: Figure out why this is needed and what it does that (h)Session
+     *       doesn't do? */
```
All keys are attached to sessions in PKCS #11. Like lifetime of the key is tied to that session, and when the session is destroyed, session keys are automatically freed.

cipherboy replied:
Ah, so is this correct understanding?

 - A `SFTKSession` is the actual "session" (with `hSession` being the numerical handle referencing a unique session).
 - Each object has its own `SFTKSessionObject` object that tracks what sessions that object is part of.
 - Handles are tied to a session?
 - So the `sftk_narrowToSessionObject` returns the `SFTKSessionObject` reference for this object+session combination (which we need to annotate that this is a derived key)
 - And then `sftk_handleObject` latter ties `hSession` / `session` to this key's `SFTKSessionObject`?

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+
+done:
+    /* NULL guard necessary as sftk_FreeObject(...) doesn't handle a NULL
+     * parameter. */
```
Probably not necessary in this function since we clearly dereference key above. We should probably check at the top of this function and error out if we think key could be NULL coming in. Including this check will cause our code scanners to mistakenly think key could be NULL and flag the dereference as an error.

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+
+CK_RV kbkdf_SaveKey(SFTKObject *key, uint8_t *key_buffer, size_t key_len) {
+    /* This is at least necessary under kbkdf_CreateKey; might as well always
+     * do it. TODO: Figure out if there's any harm in always doing it. */
```
This should be fine. The mechanism always derives CKO_SECRET_KEYs. (as opposed to CKO_PRIVATE_KEYS or CKO_PUBLIC_KEYS. The one exception is the extract case, which is CKO_DATA_OBJECT, but in that case we could adjust the result before we return it.

cipherboy replied:
This section will be rewritten, sorry -- I'm working on making the KBKDF interface match the MAC interface that I proposed above -- a `kbkdf_Counter(.... SFTKObject *key ... )` and a `kbkdf_CounterRaw(.... uint8_t *key, uint32_t key_length, ... )` -- that'll make the ~TLS~ extract implementation easier I think (no need to handle creating a fake key, passing it to the KDF, and then unwrapping and freeing it).

Then if we want a CKO_DATA_OBJECT, we can call Raw instead and I think it'll be cleaner .

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+    }
+
+    /* Else, fall back to this mapping. We know kbkdf_ValidateDerived(...)
+     * passed, so this should return non-zero. */
```
If keyType is CKK_GENERIC_SECRET, then sftk_MapKeySize may return '0['. Is there a default, (like the size of the underlying prf)? or does the mechanism require the length and fails with invalid template if it doesn't exist?

cipherboy replied:
In `NSC_DeriveKey`, we get passed a `(key, keySize)` tuple that has validated `keySize != 0` if there is a default:

https://github.com/nss-dev/nss/blob/master/lib/softoken/pkcs11c.c#L6767-L6769

---

In `kbkdf_ValidateParameters(...)`, we validate the same against each derived key. 

https://github.com/cipherboy/nss/blob/nist-800-108-kbkdf/lib/softoken/kbkdf.c#L159-L167

---

This means that for any and all keys we get passed, we should know a non-zero `keySize`. My reading of the PKCS#11 v3.0 spec says that this key must be passed and must not be `NULL`, but I could be wrong. Additionally, in the NSS PKCS#11 wrapper, there's no way to pass a `NULL` key since it gets allocated for the caller (and returned if the underlying `NSC_DeriveKey` call succeeds).

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+            PR_ASSERT(keySize > *key_buffer_index);
+            PORT_Memcpy((*key_buffer) + (*key_buffer_index), output_buffer, copy_amount);
+
+            PR_ASSERT(ret_key != NULL);
```
Good, since you have cases where ret_key could be NULL, it's good to verify that it isn't here.

cipherboy replied:
I don't think `ret_key` could actually be `NULL` here. `ret_key` gets allocated in `NSC_DeriveKey`: 

https://github.com/nss-dev/nss/blob/master/lib/softoken/pkcs11c.c#L6741-L6745

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+
+    ret = sftk_handleObject(key, session);
+    *(derived_key->phKey) = key->handle;
+    /* TODO: Figure out why we still set phKey when ret != CKR_OK? */
```
Probably shouldn't. The caller won't look at it, and it may not have a real meaning.

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+            /* Clear and free the buffer, resetting back to the initial
+             * buffer state. */
+            PORT_Memset(*key_buffer, 0, keySize);
+            free(*key_buffer);
```
Good, We need to clear all our CSPs for 1) good crypto hygeine, and 2) FIPS.

cipherboy replied:
I've switched these to ZFree, which should both zero and free.

rjrelyea commented on lib/softoken/kbkdf.c:
```diff
@@ -0,0 +1,849 @@
+    }
+
+    /* Allocate memory for the PRF output */
+    output_buffer = calloc(ctx.output_size, sizeof(uint8_t));
```
OK, I just noticed you are using calloc/malloc. NSS has 'wrapped' these for style and portability reasons (well it's been a while since NSS was ported to a non-POSIX platform, it's still part of the tradition). Should use the PORT_ versions of these functions.

cipherboy replied:
Ah right, sorry. :)

rjrelyea reviewed (commented):
Looks mostly OK, I do have some comments other then the (good jobs this looks good, which are also sprinkled in the review;).

Particularly:
1) use the PORT_ wrappers for the malloc/alloc/calloc/free functions.
2) Let's get rid of the explict HMAC test in verifying if the params are valid so that we don't need to update this code whenever we add new hash functions (the explict CMAC is fine and not including AES_MACX is fine as well).
3) Some changes that will steer us way from getting false positives in our scanning tools.

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_feedbackcounterafterfixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_feedbacknocounterfixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_feedbackcounterbeforeiter.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_feedbackcounterafteriter.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_pipelinecounterafterfixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_pipelinecounterafteriter.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_pipelinecounterbeforeiter.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf_pipelinenocounterfixed.h

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP do-fast-build.sh

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
Bug 1599603 - Implement PKCS#11 NIST SP800-108 KBKDF r=rrelyea

This implements NIST SP800-108 Counter, Feedback, and Double Pipeline
mode KDFs suitable for use in SCP03 and other protocols. These KDFs were
introduced in PKCS#11 v3.0.

Resolves: BZ#1599603

Differential Revision: https://phabricator.services.mozilla.com/D54821

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy committed:
FIXUP gtests/pk11_gtest/pk11_kbkdf.cc

Signed-off-by: Alexander Scheel <ascheel@redhat.com>

cipherboy commented:
Merged upstream! \o/
